rfcs/proposed/flow_graph_serializers/README.md

```diff
@@ -0,0 +1,613 @@
+    tbb::flow::buffer_node<token_type>{g}, handles_(n_handles)
+  {}
+  /// \brief Constructs a resource_limiter with explicitly provided handles.
+  resource_limiter(tbb::flow::graph& g, std::vector<Handle>&& handles) :
+    tbb::flow::buffer_node<token_type>{g}, handles_(std::move(handles))
+  {}
+
+  /// \brief Activate the resource_limiter by placing one token to each handle
+  /// it manages into its buffer.
+  ///
+  /// This function must be called before the flow graph is started.  It
+  /// is not automatically called by the constructor because the
+  /// resource_limiter may be placed into a container (e.g. a vector) that
+  /// can grow after the resource_limiter is constructed.  If the
+  /// resource_limiter is activated before it is placed into its final
+  /// location, then the locations of the tokens in the buffer can become
+  /// invalid, resulting in memory errors when the flow graph is executed.
+  void activate()
+  {
+    // Place tokens into the buffer.
+    for (auto& handle : handles_) {
+      tbb::flow::buffer_node<token_type>::try_put(&handle);
+    }
+  }
+
+private:
+  std::vector<Handle> handles_;
+};
+```
+
+where `Handle` represents the type of a resource handle for which tokens
+can be passed throughout the graph. With this implementation, the token
+is simply a pointer to a handle owned by the resource limiter.
+
+#### C++20 support
+
+When compiling with a C++ standard of at least C++20, the
+`resource_limiter` `Handle` template parameter can be constrained to
+model a resource-handle concept.
+
+### Resource handles
+
+#### User-defined resource handles
+
+An example of a user-defined resource handle is the `DB` handle
+discussed above. A handle, in principle, can have an arbitrary structure
+with unlimited interface, so long as ownership of the handle ultimately
+reside with the `resource_limiter`.
+
+#### `default_resource_handle`
+
+For `rl_function_node` function bodies that do not need to access
+details of the resource, a default policy can be provided:
+
+``` cpp
+resource_limiter r{g};
+```
+
+This can be useful if a third-party library supports substantial
+thread-unsafe interface and there is no obvious API that should be
+attached to the handle.
+
+### `rl_function_node` constructors
+
+We imagine the following constructors could exist
+
+``` cpp
+using namespace tbb;
+flow::graph g;
+
+// 1. Already discussed above
+rl_function_node fn1{g,
+                     std::tie(gpus, root),
+                     [](Hits const&, token<GPU>, token<ROOT>) -> Tracks { ... }};
+
+// 2. No access required to a limited resource (equivalent to flow::function_node)
+rl_function_node fn2{g, flow::unlimited, [](Hits const&) -> Tracks { ... }};
+
+// 3. Access required to a limited resource, and the user body must be serialized
+rl_function_node fn3{g,
+                     std::make_tuple(flow::serial, std::ref(db_limiter)),
+                     [](Hit const&, token<DB>) -> Tracks{ ... }};
```

Comment:
Instead of adding new `rl_xxx_node` class templates, would it make sense to add corresponding constructors directly to the relevant flow-graph nodes? I.e., in each node type that supports this functionality, add  constructors with `size_t concurrency` replaced by a tuple of `resource_limiter`s (1 and 3 above). Constructor 2 is already available in the current node type.

Reply:
That is my question/desire as well. But I think it should probably be the job of the oneTBB team to evolve the proposal into this direction. For the purpose of initial acceptance, it should be fine to add this as a design option or an open question.

Reply:
We agree with both of you.  Please let us know how we can help.

Reply:
@marcpaterno @knoepfel
> Please let us know how we can help.

For now could you please just add some note in the doc that there are two possible approaches: adding new "resource limited" types of functional nodes, like proposed, and trying to extend the existing functional nodes with the necessary support.

A deeper discussion of pros and cons etc., like in Aleksei's comment above, should likely be driven by someone from the team; of course you will be welcome to participate.

Reply:
@akukanov, we've included the alternative proposal (i.e. just adding constructors to existing nodes) in the proposal description above.